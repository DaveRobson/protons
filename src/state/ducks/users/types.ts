const REQUEST_USERS = "REQUEST_USERS";
const REQUEST_USERS_PENDING = "REQUEST_USERS_PENDING";
const REQUEST_USERS_SUCCESS = "REQUEST_USERS_SUCCESS";
const REQUEST_USERS_FAILURE = "REQUEST_USERS_FAILURE";

export interface User {
  firstname: string;
  lastname: string;
  email: string;
}

export {
  REQUEST_USERS,
  REQUEST_USERS_PENDING,
  REQUEST_USERS_SUCCESS,
  REQUEST_USERS_FAILURE
};
