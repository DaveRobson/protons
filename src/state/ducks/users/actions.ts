import * as types from "./types";
import { User } from "./types";
import { Action } from "redux";

export interface RequestSuccessAction extends Action {
  data: User[];
}

interface ErrorAction extends Action {
  message: string;
}

const requestUsers = (): Action => ({
  type: types.REQUEST_USERS
});

const requestUsersPending = (): Action => ({
  type: types.REQUEST_USERS_PENDING
});

const requestUsersSuccess = (data: User[]): RequestSuccessAction => ({
  type: types.REQUEST_USERS_SUCCESS,
  data
});

const requestUsersFailure = (message: string): ErrorAction => ({
  type: types.REQUEST_USERS_FAILURE,
  message
});

export {
  requestUsers,
  requestUsersSuccess,
  requestUsersPending,
  requestUsersFailure
};
