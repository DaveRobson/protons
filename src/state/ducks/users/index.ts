import usersReducer from "./reducers";
import usersEpic from "./epic";

export { usersReducer, usersEpic };
