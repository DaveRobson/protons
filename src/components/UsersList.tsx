import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { requestUsers } from "../state/ducks/users/actions";
import { User } from "../state/ducks/users/types";

const userItem = (user: User): React.ReactElement => (
  <div key={user.email} id="details">
    <label>Firstname: {user.firstname}</label>
    <br />
    <label>Lastname: {user.lastname}</label>
    <br />
    <label>Email: {user.email}</label>
  </div>
);

const UsersList: React.FC = (): React.ReactElement => {
  const users = useSelector(
    (state: { users: { data: User[]; loading: boolean } }) => state.users
  );
  const dispatch = useDispatch();
  const onClick = (): void => {
    dispatch(requestUsers());
  };

  return (
    <>
      <button onClick={onClick}>Get Users</button>
      <br />
      {users.loading ? <div>...loading</div> : users.data.map(userItem)}
    </>
  );
};

export default UsersList;
