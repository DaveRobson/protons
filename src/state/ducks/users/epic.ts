import { ofType } from "redux-observable"
import { REQUEST_USERS, User } from "./types"
import { map, mergeMap } from "rxjs/operators"
import { db } from "../../../config/firebase";
import { from, Observable } from "rxjs"
import { requestUsersSuccess } from "./actions"
import { Action } from "redux"
import firebase from "firebase";
import QueryDocumentSnapshot = firebase.firestore.QueryDocumentSnapshot
import DocumentData = firebase.firestore.DocumentData

const firebaseUserMapper = (input: QueryDocumentSnapshot<DocumentData>): User => {
  return {
    firstname: input.get("firstname"),
    lastname: input.get("lastname"),
    email: input.get("email")
  }
}

export default (action$: Observable<Action>) => action$.pipe(
  ofType(REQUEST_USERS),
  mergeMap(action =>
    from(db.collection("users").get()).pipe(
      map(response => requestUsersSuccess(response.docs.map(firebaseUserMapper)))
    )
  )
);