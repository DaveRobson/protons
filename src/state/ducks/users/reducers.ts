import { REQUEST_USERS, REQUEST_USERS_SUCCESS, User } from "./types";
import { RequestSuccessAction } from "./actions";

export default (
  state: { data: User[]; loading: boolean } = { data: [], loading: false },
  action: RequestSuccessAction
): { data: User[]; loading: boolean } => {
  switch (action.type) {
    case REQUEST_USERS:
      return {
        ...state,
        loading: true
      };
    case REQUEST_USERS_SUCCESS:
      return {
        ...state,
        data: action.data,
        loading: false
      };
    default:
      return state;
  }
};
