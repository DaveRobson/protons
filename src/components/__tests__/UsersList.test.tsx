import React from 'react';
import {combineReducers, createStore} from "redux";
import usersReducer from "../../state/ducks/users/reducers";
import {Provider} from "react-redux";
import UsersList from "../UsersList";
import {requestUsers} from "../../state/ducks/users/actions";
import { mount } from "enzyme";

describe('<UsersList /> unit test', () => {
    const initialState = {users: {data: [], loading: false}};
    const reducers = combineReducers({users: usersReducer});
    const getWrapper = ((mockStore = createStore(reducers, initialState)) => mount(
        <Provider store={mockStore}>
            <UsersList />
        </Provider>
    ));

    it('should dispatch user request', () => {
        const mockStore = createStore(reducers, initialState);
        mockStore.dispatch = jest.fn();
        const wrapper = getWrapper(mockStore);
        expect(wrapper.find("#details")).toEqual({});
        wrapper.find('button').simulate('click');
        expect(mockStore.dispatch).toHaveBeenCalledWith(requestUsers());
    });

    it('should render user list items', () => {
        const mockStore = createStore(reducers, {
            ...initialState,
            users: {
                ...initialState.users,
                data: [{firstname: "David", lastname: "robson", email: "david@davidrobson.co.uk"}]
            }
        });
        mockStore.dispatch = jest.fn();
        const wrapper = getWrapper(mockStore);

        expect(wrapper.find(UsersList)).toMatchSnapshot();

    });
});