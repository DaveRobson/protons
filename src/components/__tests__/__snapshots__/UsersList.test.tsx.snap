// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersList /> unit test should render user list items 1`] = `
<UsersList>
  <button
    onClick={[Function]}
  >
    Get Users
  </button>
  <br />
  <div
    id="details"
    key="david@davidrobson.co.uk"
  >
    <label>
      Firstname: 
      David
    </label>
    <br />
    <label>
      Lastname: 
      robson
    </label>
    <br />
    <label>
      Email: 
      david@davidrobson.co.uk
    </label>
  </div>
</UsersList>
`;
